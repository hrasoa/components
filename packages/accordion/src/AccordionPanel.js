// @flow
import React, { Component } from 'react';
import disabled from 'ally.js/maintain/disabled';

type Props = {
  id: string,
  expandedClass?: string,
  expanded?: boolean,
  addPanel: (
    id: string,
    ref: PanelRef,
    isInitiallyExpanded: ?boolean,
  ) => void,
  className?: ?string,
  isExpanded?: boolean,
  disableInnert?: boolean,
};

class AccordionPanel extends Component<Props> {
  static defaultProps = {
    expandedClass: 'is-expanded',
  }

  constructor(props: Props) {
    super(props);
    this.handleInnert = null;
    this.props.addPanel(this.props.id, this.ref, this.props.expanded);
  }

  componentDidMount() {
    if (this.ref.current && !this.props.disableInnert) {
      this.handleInnert = this.props.isExpanded ? null : disabled({
        context: this.ref.current,
      });
    }
  }

  shouldComponentUpdate(nextProps: Props) {
    return this.props.isExpanded !== nextProps.isExpanded;
  }

  componentDidUpdate() {
    if (this.props.disableInnert) return;

    if (this.props.isExpanded && this.handleInnert) {
      this.handleInnert.disengage();
    } else if (this.ref.current) {
      this.handleInnert = disabled({
        context: this.ref.current,
      });
    }
  }

  componentWillUnmount() {
    if (this.handleInnert) {
      this.handleInnert.disengage();
      this.handleInnert = null;
    }
  }

  handleInnert: null | { disengage: () => void };

  ref: PanelRef = React.createRef();

  render() {
    const {
      className,
      expandedClass,
      isExpanded,
      expanded,
      addPanel,
      ...rest
    } = this.props;
    return (
      <dd
        className={getClassName(className, expandedClass, isExpanded)}
        ref={this.ref}
        {...rest}
      />
    );
  }
}

function getClassName(className, expandedClass, isExpanded): ?string {
  const name = [
    className || '',
    isExpanded ? expandedClass : '',
  ].join(' ').trim();
  return name.length ? name : null;
}

export default AccordionPanel;
