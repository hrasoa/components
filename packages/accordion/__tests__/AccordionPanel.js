// @flow
import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import AccordionPanel from '../src/AccordionPanel';

const mockAddPanel = jest.fn();

beforeEach(() => {
  mockAddPanel.mockClear();
});

describe('AccordionPanel', () => {
  test('initial', () => {
    const component = renderer.create(
      <AccordionPanel
        id="panel-1"
        addPanel={mockAddPanel}
      >
        Panel 1
      </AccordionPanel>,
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('is expanded', () => {
    const component = renderer.create(
      <AccordionPanel
        id="panel-1"
        className="accordion__panel"
        aria-labelledby="header-1"
        role="region"
        isExpanded
        addPanel={mockAddPanel}
      >
        Panel 1
      </AccordionPanel>,
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('addPanel', () => {
    const wrapper = mount(
      <AccordionPanel
        id="panel-1"
        addPanel={mockAddPanel}
      >
        Panel 1
      </AccordionPanel>,
    );
    expect(mockAddPanel).toHaveBeenCalledTimes(1);
    expect(mockAddPanel).toHaveBeenCalledWith(
      'panel-1',
      { current: wrapper.find('dd').first().instance() },
      undefined,
    );
  });

  test('addPanel initiallyExpanded', () => {
    const wrapper = mount(
      <AccordionPanel
        id="panel-1"
        addPanel={mockAddPanel}
        expanded
      >
        Panel 1
      </AccordionPanel>,
    );
    expect(mockAddPanel).toHaveBeenCalledTimes(1);
    expect(mockAddPanel).toHaveBeenCalledWith(
      'panel-1',
      { current: wrapper.find('dd').first().instance() },
      true,
    );
  });
});
