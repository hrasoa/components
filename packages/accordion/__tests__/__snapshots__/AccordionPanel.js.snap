// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccordionPanel Snapshot testing expanded state 1`] = `
<dd
  className="is-expanded"
  id="panel-1"
>
  Panel 1
</dd>
`;

exports[`AccordionPanel Snapshot testing minimal 1`] = `
<dd
  className={null}
  id="panel-1"
>
  Panel 1
</dd>
`;

exports[`AccordionPanel Snapshot testing with more props 1`] = `
<dd
  aria-labelledby="header-1"
  className="accordion__panel"
  id="panel-1"
  role="region"
  style={
    Object {
      "color": "inherit",
    }
  }
>
  Panel 1
</dd>
`;
